File: DynamicProgramming/Ruby/grid_traveller.rb
# Grid traveller algorithm using memoization
# Time complexity O(m * n), Space complexity O(m + n)
def grid_traveler(m, n, memo = {})
    checker = "#{m},#{n}"
    return memo[checker] if memo[checker]
    return 1 if(m == 1 && n == 1)
    return 0 if(m == 0 || n == 0)

    memo[checker] = grid_traveler(m - 1, n, memo) + grid_traveler(m, n - 1, memo)
end

puts grid_traveler(2, 3)
puts grid_traveler(3, 3)
puts grid_traveler(4, 2)
puts grid_traveler(6, 5)
puts grid_traveler(18, 18)

File: DynamicProgramming/JavaScript/gridTraveller.js
const gridTraveler = (m, n, memo = {}) => {
    const key = m + '_' + n;
    if(key in memo) return memo[key];
    if(m == 0 || n == 0) return 0;
    if(m ==1 && n == 1) return 1;
    
    memo[key] = gridTraveler(m - 1, n, memo) + gridTraveler(m, n - 1, memo);

    return memo[key];
};

console.log(gridTraveler(2, 3));
console.log(gridTraveler(3, 3));
console.log(gridTraveler(4, 4));
console.log(gridTraveler(5, 10));
console.log(gridTraveler(20, 20));
